Build problems with LLVM and Enzyme

Hello Enzyme. I want to build Enzyme with  llvm-project from source, and just meet some problems.

From the enzyme-mlir.yml, I checkout the llvm-project to cbc378ecb87e the same as Polygeist now.
```
r@b:~/Desktop/all_project/llvm-project$ git log --oneline | head -n 1
cbc378ecb87e GlobalISel: Artifact combine merge-like and unmerges into merge-like
```

and follow the step in .yml's  MLIR build to enable projects with "llvm;clang;mlir;openmp".

then Enzyme related command is 
```
cmake .. -DCMAKE_BUILD_TYPE=Release -DLLVM_DIR=/home/r/Desktop/all_project/llvm-project/build/lib/cmake/llvm  -DLLVM_EXTERNAL_LIT=`which lit`
```

get unexpected error 
```
/home/r/Desktop/workSpace/Enzyme/enzyme/Enzyme/ActivityAnalysis.cpp: In lambda function:
/home/r/Desktop/workSpace/Enzyme/enzyme/Enzyme/ActivityAnalysis.cpp:1769:22: error: ‘class llvm::AAResults’ has no member named ‘getModRefBehavior’
 1769 |           AARes = AA.getModRefBehavior(CB).getModRef();
```

Meanwhie  to search keyword **getModRefBehavior** in the text editor, there is exactly empty result.

I wonder is that the commitID of llvm-project   not  cbc378ecb87e ?

--- a/enzyme/Enzyme/ActivityAnalysis.cpp
+++ b/enzyme/Enzyme/ActivityAnalysis.cpp
@@ -1766,7 +1766,7 @@ bool ActivityAnalyzer::isConstantValue(TypeResults const &TR, Value *Val) {
       if (!memval->getType()->isPointerTy()) {
         if (auto CB = dyn_cast<CallInst>(I)) {
 #if LLVM_VERSION_MAJOR >= 16
-          AARes = AA.getModRefBehavior(CB).getModRef();
+          AARes = AA.getMemoryEffects(CB).getModRef();
 #else
           AARes = createModRefInfo(AA.getModRefBehavior(CB));
 #endif

--- a/enzyme/Enzyme/MLIR/Dialect/EnzymeOps.td
+++ b/enzyme/Enzyme/MLIR/Dialect/EnzymeOps.td
@@ -36,7 +36,7 @@ def ActivityArrayAttr : TypedArrayAttrBase<
     ActivityAttr, "Array of argument activity states">;
 
 def PlaceholderOp : Enzyme_Op<"placeholder",
-    [NoSideEffect]> {
+    [Pure]> {
   let results = (outs AnyType:$output);
 }
 

--- a/enzyme/Enzyme/MLIR/Interfaces/GradientUtils.cpp
+++ b/enzyme/Enzyme/MLIR/Interfaces/GradientUtils.cpp
@@ -216,7 +216,7 @@ void mlir::enzyme::MGradientUtils::forceAugmentedReturns() {
   // getContext(cast<BasicBlock>(getNewFromOriginal(&oBB)), loopContext);
 
   oldFunc.walk([&](Block *blk) {
-    if (blk == &oldFunc.getBody().getBlocks().front())
+    if (blk == &oldFunc.getFunctionBody().getBlocks().front())
       return;
     auto nblk = getNewFromOriginal(blk);
     for (auto val : llvm::reverse(blk->getArguments())) {
@@ -510,7 +510,7 @@ FunctionOpInterface CloneFunctionWithReturns(
     DIFFE_TYPE returnType, Twine name, BlockAndValueMapping &VMap,
     std::map<Operation *, Operation *> &OpMap, bool diffeReturnArg,
     mlir::Type additionalArg) {
-  assert(!F.getBody().empty());
+  assert(!F.getFunctionBody().empty());
   // F = preprocessForClone(F, mode);
   // llvm::ValueToValueMapTy VMap;
   auto FTy = getFunctionTypeForClone(
@@ -518,7 +518,7 @@ FunctionOpInterface CloneFunctionWithReturns(
       additionalArg, constant_args, diffeReturnArg, returnValue, returnType);
 
   /*
-  for (Block &BB : F.getBody().getBlocks()) {
+  for (Block &BB : F.getFunctionBody().getBlocks()) {
     if (auto ri = dyn_cast<ReturnInst>(BB.getTerminator())) {
       if (auto rv = ri->getReturnValue()) {
         returnvals.insert(rv);
@@ -538,12 +538,12 @@ FunctionOpInterface CloneFunctionWithReturns(
   table.insert(NewF);
   SymbolTable::setSymbolVisibility(NewF, SymbolTable::Visibility::Private);
 
-  cloneInto(&F.getBody(), &NewF.getBody(), VMap, OpMap);
+  cloneInto(&F.getFunctionBody(), &NewF.getFunctionBody(), VMap, OpMap);
 
   {
-    auto &blk = NewF.getBody().front();
+    auto &blk = NewF.getFunctionBody().front();
     for (ssize_t i = constant_args.size() - 1; i >= 0; i--) {
-      mlir::Value oval = F.getBody().front().getArgument(i);
+      mlir::Value oval = F.getFunctionBody().front().getArgument(i);
       if (constant_args[i] == DIFFE_TYPE::CONSTANT)
         constants.insert(oval);
       else
@@ -771,7 +771,7 @@ FunctionOpInterface mlir::enzyme::MEnzymeLogic::CreateForwardDiff(
     std::vector<DIFFE_TYPE> constants, MTypeAnalysis &TA, bool returnUsed,
     DerivativeMode mode, bool freeMemory, size_t width, mlir::Type addedType,
     MFnTypeInfo type_args, std::vector<bool> volatile_args, void *augmented) {
-  if (fn.getBody().empty()) {
+  if (fn.getFunctionBody().empty()) {
     llvm::errs() << fn << "\n";
     llvm_unreachable("Differentiating empty function");
   }
@@ -829,7 +829,7 @@ FunctionOpInterface mlir::enzyme::MEnzymeLogic::CreateForwardDiff(
                                   unnecessaryInstructions, gutils, TLI);
                                   */
 
-  for (Block &oBB : gutils->oldFunc.getBody().getBlocks()) {
+  for (Block &oBB : gutils->oldFunc.getFunctionBody().getBlocks()) {
     // Don't create derivatives for code that results in termination
     if (guaranteedUnreachable.find(&oBB) != guaranteedUnreachable.end()) {
       auto newBB = gutils->getNewFromOriginal(&oBB);
@@ -865,7 +865,7 @@ FunctionOpInterface mlir::enzyme::MEnzymeLogic::CreateForwardDiff(
 
   // gutils->eraseFictiousPHIs();
 
-  mlir::Block *entry = &gutils->newFunc.getBody().front();
+  mlir::Block *entry = &gutils->newFunc.getFunctionBody().front();
 
   // cleanupInversionAllocs(gutils, entry);
   // clearFunctionAttributes(gutils->newFunc);

--- a/enzyme/Enzyme/MLIR/Passes/EnzymeMLIRPass.cpp
+++ b/enzyme/Enzyme/MLIR/Passes/EnzymeMLIRPass.cpp
@@ -73,7 +73,7 @@ struct DifferentiatePass : public DifferentiatePassBase<DifferentiatePass> {
     size_t width = 1;
 
     std::vector<bool> volatile_args;
-    for (auto &a : fn.getBody().getArguments()) {
+    for (auto &a : fn.getFunctionBody().getArguments()) {
       volatile_args.push_back(!(mode == DerivativeMode::ReverseModeCombined));
     }
 

